File: benches/lang-comparison/c/quicksort.c
// Quicksort with array partitioning
#include <stdio.h>
#include <stdint.h>

void swap(int64_t *arr, int i, int j) {
    int64_t t = arr[i];
    arr[i] = arr[j];
    arr[j] = t;
}

int partition(int64_t *arr, int lo, int hi) {
    int64_t pivot = arr[hi];
    int i = lo;
    for (int j = lo; j < hi; j++) {
        if (arr[j] <= pivot) {
            swap(arr, i, j);
            i++;
        }
    }
    swap(arr, i, hi);
    return i;
}

void quicksort(int64_t *arr, int lo, int hi) {
    if (lo < hi) {
        int p = partition(arr, lo, hi);
        quicksort(arr, lo, p - 1);
        quicksort(arr, p + 1, hi);
    }
}

int main() {
    int64_t arr[] = {38, 27, 43, 3, 9, 82, 10, 55, 1, 77};
    quicksort(arr, 0, 9);
    for (int i = 0; i < 10; i++) {
        printf("%lld\n", arr[i]);
    }
    return 0;
}

File: benches/lang-comparison/rust/quicksort.rs
// Quicksort with array partitioning
fn partition(arr: &mut [i64], lo: usize, hi: usize) -> usize {
    let pivot = arr[hi];
    let mut i = lo;
    for j in lo..hi {
        if arr[j] <= pivot {
            arr.swap(i, j);
            i += 1;
        }
    }
    arr.swap(i, hi);
    i
}

fn quicksort(arr: &mut [i64], lo: isize, hi: isize) {
    if lo < hi {
        let p = partition(arr, lo as usize, hi as usize);
        quicksort(arr, lo, p as isize - 1);
        quicksort(arr, p as isize + 1, hi);
    }
}

fn main() {
    let mut arr = [38i64, 27, 43, 3, 9, 82, 10, 55, 1, 77];
    let len = arr.len() as isize;
    quicksort(&mut arr, 0, len - 1);
    for val in &arr {
        println!("{}", val);
    }
}
